packages/nextjs/src/middlewareClient.ts

```diff
@@ -102,3 +114,228 @@ export function createMiddlewareClient<
+		// note how this does not call super.setItem
+		// this is intentional, as that implementation is used from commitCookies
+
+		this.items[key] = value;
+		delete this.deleteItems[key];
+	}
+
+	async removeItem(key: string): Promise<void> {
+		// note how this does not call super.removeItem
+		// this is intentional, as that implementation is used from commitCookies
+
+		delete this.items[key];
+		this.deleteItems[key] = true;
+	}
+
+	commitCookies(res: NextResponse) {
+		try {
+			this.res = res;
+
+			const supRemoveItem = super.removeItem.bind(this);
+			const supSetItem = super.setItem.bind(this);
+
+			Object.keys(this.deleteItems).forEach((name) => {
+				supRemoveItem(name);
+			});
+
+			Object.entries(this.items).forEach(([name, value]) => {
+				supSetItem(name, value);
+			});
+		} finally {
+			this.res = null;
+		}
+	}
+
+	/**
+	 * Only used initially from {@link #getItem}.
+	 */
+	protected getCookie(name: string): string | null | undefined {
+		const cookies = parseCookies(this.req.headers.get('cookie') ?? '');
+		return cookies[name];
+	}
+
+	/**
+	 * Only called from {@link #commitCookies}.
+	 */
+	protected setCookie(name: string, value: string): void {
+		return this._setCookie(name, value);
+	}
+
+	/**
+	 * Only called from {@link #commitCookies}.
+	 */
+	protected deleteCookie(name: string): void {
+		this._setCookie(name, '', {
+			maxAge: 0
+		});
+	}
+
+	private _setCookie(name: string, value: string, options?: DefaultCookieOptions): void {
+		const newSessionStr = serializeCookie(name, value, {
+			...this.cookieOptions,
+			...options,
+			// Allow supabase-js on the client to read the cookie as well
+			httpOnly: false
+		});
+
+		if (this.res?.headers) {
+			this.res.headers.append('set-cookie', newSessionStr);
+		}
+	}
+}
+
+/**
+ * Returns an array of SupabaseClient and a function that when called produces
+ * the {@link NextResponse#next()} response to be returned in the NextJS
+ * middleware function.
+ *
+ * You must call this function towards the end of your middleware function,
+ * typically as part of the return statement. Failing to return the result from
+ * this function can result in your users being randomly logged out.
+ *
+ * @example Here's a basic example:
+ * ```
+ * export function middleware(request: NextRequest) {
+ *   const [supabaseClient, nextResponse] = createMiddlewareClient2(request, {})
+ *
+ *   const { data: { user } } = await supabaseClient.auth.getUser()
+ *
+ *   if (!user) {
+ *     // there's no user session, ask them to log in
+ *     return NextResponse.redirect(new URL('/sign-in-page-in-your-app'))
+ *   }
+ *
+ *   return nextResponse()
+ * }
+ * ```
+ */
+export function createMiddlewareClient2<
+	Database = any,
+	SchemaName extends string & keyof Database = 'public' extends keyof Database
+		? 'public'
+		: string & keyof Database,
+	Schema extends GenericSchema = Database[SchemaName] extends GenericSchema
+		? Database[SchemaName]
+		: any
+>(
+	req: NextRequest,
```

Comment:
hmm it seems like the only difference between `createMiddlewareClient` and `createMiddlewareClient2`  in terms of the input params is this argument (`createMiddlewareClient` takes in a `context: { req: NextRequest; res: NextResponse }` instead 

which begs the question - why can't we just modify `createMiddlewareClient` to use this new implementation - where we set the cookie on the request after cookie has been updated with the new session?

Reply:
I think it's doable, but I want to signal to users that they need to move to a different way of use. A new method name seems to signal this more effectively than signature/type change, though that's also an option.

We need to keep both ways to prevent complete compile-time breakage (even though functionality is broken) to not disrupt folks unnecessarily.

Reply:
or we could change the functionality, bump the major version of the package and indicate in the changelog that the underlying implementation of `createMiddlewareClient` is different

Reply:
I don't want to bump the major version because nobody will upgrade to that, due to `^`. The old method remains just so they are able to compile, it will show up crossed-out in VScode and will add a postinstall log message.